rakelib/docs.rake

```diff
@@ -196,25 +224,69 @@ begin
       $stdout.puts "Generated bolt-defaults.yaml reference at:\n\t#{filepath}"
     end
 
+    desc "Generate markdown docs for Bolt's data types"
+    task type_reference: :generate_strings do
+      filepath = File.expand_path('../documentation/bolt_types_reference.md', __dir__)
+      template = File.expand_path('../documentation/templates/bolt_types_reference.md.erb', __dir__)
+
+      @types = @puppet_strings['data_types'].map do |data|
+        functions = []
+
+        # Turn types that include TypeReference into just the name of the
+        # type.
+        data = Bolt::Util.walk_vals(data) do |val|
+          if val.is_a?(String)
+            val.gsub(/TypeReference\['(\w*)'\]/, '\1')
+          else
+            val
+          end
+        end
+
+        # Add attributes, which are listed under the type's params
+        data.dig('docstring', 'tags').each do |param|
+          functions << {
+            'name' => param['name'],
+            'desc' => param['text'].tr("\n", ' '),
+            'type' => param['types'].first
+          }
+        end
+
+        # Add functions
+        data['functions'].each do |func|
+          if data['name'] == 'ResultSet' && func['name'] == '[]'
+            type = 'Variant[Result, ApplyResult, Array[Variant[Result, ApplyResult]]]'
+          else
+            type = func.dig('docstring', 'tags').first['types']
```

Comment:
There are a number of other cases where the first type is the type of a parameter to the function, not the return type. I think it's safest to get `.last` here instead and verify that the `tag_name` is `return`.